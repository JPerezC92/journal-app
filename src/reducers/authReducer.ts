import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface AuthState {
  isLoggedIn: boolean;
  displayName: string;
  uuid: string;
}

const initialState: AuthState = {
  isLoggedIn: false,
} as AuthState;

export const authSlice = createSlice({
  name: "[Auth]",
  initialState,
  reducers: {
    login: (_, action: PayloadAction<AuthState>): AuthState => action.payload,
    logout: () => initialState,
  },
});

export const { login, logout } = authSlice.actions;

const authReducer = authSlice.reducer;

export default authReducer;
