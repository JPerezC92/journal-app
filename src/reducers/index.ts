export * from "./authReducer";
export * from "./uiReducer";
export * from "./notesReducer";
