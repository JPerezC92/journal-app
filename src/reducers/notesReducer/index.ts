export * from "./notesReducer";
export * from "./notesThunks";
