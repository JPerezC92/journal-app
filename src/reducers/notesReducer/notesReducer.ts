import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface Note {
  id: string;
  title: string;
  body: string;
  imageUrl: string | null;
  date: number;
}

interface NotesState {
  notes: Note[];
  active: Note | null;
}
const initialState: NotesState = { notes: [], active: null };

const notesSlice = createSlice({
  name: "[Notes]",
  initialState,
  reducers: {
    addNewNote: (state, { payload }: PayloadAction<Note>) => {
      state.notes = [...state.notes, payload];
    },
    setNoteActive: (state, action: PayloadAction<Partial<Note>>) => {
      state.active = { ...state.active!, ...action.payload };
    },
    setNotes: (state, action: PayloadAction<Note[]>) => {
      state.notes = action.payload;
    },
    refreshNote: (state, { payload }: PayloadAction<Note>) => {
      state.notes = state.notes.map((note) =>
        note.id === payload.id ? payload : note
      );
    },

    deleteNote: (state, { payload }: PayloadAction<string>) => {
      state.notes = state.notes.filter((note) => note.id !== payload);
      state.active = null;
    },
    notesLogoutCleaning: (state) => {
      state.notes = [];
      state.active = null;
    },
  },
});

export const notesActions = notesSlice.actions;

export const notesReducer = notesSlice.reducer;
