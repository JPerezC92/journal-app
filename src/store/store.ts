import { configureStore } from "@reduxjs/toolkit";
import { uiReducer, authReducer, notesReducer } from "../reducers";

export const reducers = {
  authReducer,
  uiReducer,
  notesReducer,
};

export const store = configureStore({
  reducer: reducers,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
