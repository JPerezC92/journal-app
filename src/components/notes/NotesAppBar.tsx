import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploadingImg } from "../../reducers";
import { RootState } from "../../store/store";

const NotesAppBar = () => {
  const { active } = useSelector((state: RootState) => state.notesReducer);

  const dispatch = useDispatch();

  const handleSave = () => {
    if (active) dispatch(startSaveNote(active));
  };

  const handleUploadImage = () => {
    document.getElementById("imageSelector")?.click();
  };

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files![0] || null;

    if (file) dispatch(startUploadingImg(file));
  };

  return (
    <div className="notes__appbar">
      <span>28 Agosto 2021</span>

      <input
        type="file"
        name="imageSelector"
        id="imageSelector"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />

      <div>
        <button className="btn" onClick={handleUploadImage} type="button">
          Picture
        </button>
        <button className="btn" onClick={handleSave} type="button">
          Save
        </button>
      </div>
    </div>
  );
};

export default NotesAppBar;
