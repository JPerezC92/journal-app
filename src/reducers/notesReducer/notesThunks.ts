import { createAsyncThunk } from "@reduxjs/toolkit";
import { db } from "../../firebase";
import { RootState } from "../../store/store";
import { Note, notesActions } from "./notesReducer";

enum NotesThunks {
  START_NEW_NOTE = "START_NEW_NOTE",
}

export const startNewNote = createAsyncThunk(
  NotesThunks.START_NEW_NOTE,
  async (_, { dispatch, getState }) => {
    const { uid } = (getState() as RootState).authReducer.user;
    const newNote: Note = {
      id: null,
      title: "",
      body: "",
      date: new Date().getTime(),
      imageUrl: null,
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(notesActions.setNoteActive({ id: docRef.id, note: newNote }));
  }
);
