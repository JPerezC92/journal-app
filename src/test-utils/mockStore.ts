import {
  AnyAction,
  CombinedState,
  combineReducers,
  configureStore,
  createSlice,
} from "@reduxjs/toolkit";
import { reducers } from "../store/store";

const storeCleanSlice = createSlice({
  name: "store",
  initialState: {},
  reducers: {
    clean: () => {},
  },
});

export const storeCleanActions = storeCleanSlice.actions;
const storeCleanReducer = storeCleanSlice.reducer;
const combinedReducers = combineReducers({ ...reducers, storeCleanReducer });

const testReducers = (
  state: CombinedState<any> | undefined,
  action: AnyAction
) => {
  if (action.type === "store/clean") {
    state = undefined;
  }
  return combinedReducers(state, action);
};

export const mockStore = configureStore({
  reducer: testReducers,
});

export type MockRootState = ReturnType<typeof mockStore.getState>;
export type MockAppDispatch = typeof mockStore.dispatch;
