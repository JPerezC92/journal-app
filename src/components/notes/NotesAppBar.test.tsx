import { act, fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { Note, notesActions } from "../../reducers";
import * as notesThunks from "../../reducers/notesReducer/notesThunks";
import { store } from "../../store/store";
import { mockStore } from "../../test-utils/mockStore";
import NotesAppBar from "./NotesAppBar";

const noteMock: Note = {
  id: "1",
  title: "Test title",
  body: "Test body",
  date: new Date().getDate(),
  imageUrl:
    "https://res.cloudinary.com/jperezc92/image/upload/v1627759301/ezxhoydkvqaidpcf6loq.png",
};

describe("Test on <NotesAppBar />", () => {
  afterEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();
    mockStore.dispatch(notesActions.setNoteActive(noteMock));
  });

  test("should match the snapshot correctly", () => {
    store.dispatch(notesActions.setNoteActive(noteMock));

    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <NotesAppBar />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should call startSaveNote", () => {
    jest.spyOn(notesThunks, "startSaveNote");

    const component = render(
      <Provider store={mockStore}>
        <MemoryRouter>
          <NotesAppBar />
        </MemoryRouter>
      </Provider>
    );

    const saveButton = component.getByText("Save");

    act(() => {
      fireEvent.click(saveButton);
    });

    expect(notesThunks.startSaveNote).toHaveBeenCalledTimes(1);
    expect(notesThunks.startSaveNote).toHaveBeenCalledWith(noteMock);
  });

  test("should call startUploadingImg", () => {
    jest.spyOn(notesThunks, "startUploadingImg");

    const component = render(
      <Provider store={mockStore}>
        <MemoryRouter>
          <NotesAppBar />
        </MemoryRouter>
      </Provider>
    );

    const pictureButton = component.getByText("Picture");
    const fileInput = component.getByTestId("fileInput");

    const file = new File([], "foto.png");

    fireEvent.change(fileInput, { target: { files: [file] } });

    act(() => {
      fireEvent.click(pictureButton);
    });

    expect(notesThunks.startUploadingImg).toHaveBeenCalledTimes(1);
    expect(notesThunks.startUploadingImg).toHaveBeenCalledWith(file);
  });
});
