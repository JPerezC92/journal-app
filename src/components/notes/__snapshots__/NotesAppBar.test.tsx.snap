// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <NotesAppBar /> should match the snapshot correctly 1`] = `
<div
  className="notes__appbar"
>
  <span>
    Dec 31, 1969
  </span>
  <input
    data-testid="fileInput"
    id="imageSelector"
    name="imageSelector"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      Picture
    </button>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
  </div>
</div>
`;
